### PR TITLE
Changes for registering for OVSDB Port table updates

### PR DESCRIPTION
**Pull request by rparulek**
Changes for registering for OVSDB Port table updates

### CODE REVIEW

File: api/Port.go [hunk 1]
```diff
@@ -7,6 +7,49 @@ import (
...
 
+// Empty structure.
+type Empty struct{}
+
+// OvsdbEventType defines event type sent to OVSDB
+type OvsdbEventType string
+
+// Const definitions for OVSDB event types
+const (
+    Add    OvsdbEventType = "ADD"
```
Comment (vishpat) [inline api/Port.go]: Remove

File: api/Port.go [hunk 2]
```diff
@@ -7,6 +7,49 @@ import (
...
+    Update OvsdbEventType = "UPDATE"
+    Delete OvsdbEventType = "DELETE"
```
Comment (vishpat) [inline api/Port.go]: remove

File: api/Port.go [hunk 3]
```diff
@@ -7,6 +7,49 @@ import (
...
+type OvsdbClient struct {
+    mConn         *libovsdb.OvsdbClient
+    mUpdateChan   chan *libovsdb.TableUpdates
+    StopChannel   chan bool
+    ClientChannel chan ClientEvent
+}
+
+// PortState defines details to be populated
+// for container port resolved in OVSDB
+type PortState struct {
```
Comment (vishpat) [inline api/Port.go]: Not if state is the right identifier. I would rename to PortIPv4Info

File: api/Port.go [hunk 4]
```diff
@@ -151,3 +194,168 @@ func (vrsConnection *VRSConnection) UpdatePortMetadata(name string, metadata map
...
+
+// GetNuagePortTableUpdate will register with OVSDB
+// for Nuage Port table updates and return as soon as
+// port table entry gets populated
+func (ovsdbc *OvsdbClient) GetNuagePortTableUpdate() <-chan ClientEvent {
+    var err error
+    var conn *libovsdb.OvsdbClient
+    ovsdbc.mUpdateChan = make(chan *libovsdb.TableUpdates)
+    clientChan := make(chan ClientEvent, 1)
+    for {
```
Comment (vishpat) [inline api/Port.go]: Get rid of this. This method needs to be called on the VRSConnection object. Please re-use the ovsdb connection in that object.

Comment (vishpat) [reply api/Port.go]: Also don't connect using tcp

File: api/Port.go [hunk 5]
```diff
@@ -151,3 +194,168 @@ func (vrsConnection *VRSConnection) UpdatePortMetadata(name string, metadata map
...
+            log.Infof("Couldn't connect to ovsdb server while trying to register for Nuage Port table updates")
+        } else {
+            break
+        }
+        time.Sleep(time.Second * time.Duration(5))
+    }
+    log.Infof("Started ovsdb client for monitoring")
+
+    ovsdbc.mConn = conn
+    ovsdbc.mConn.Register(ovsdbc)
```
Comment (vishpat) [inline api/Port.go]: Change to conn.Register(ovsdbc) to make it more clearer

File: api/Port.go [hunk 6]
```diff
@@ -151,3 +194,168 @@ func (vrsConnection *VRSConnection) UpdatePortMetadata(name string, metadata map
...
+    }
+    for table, tableSchema := range schema.Tables {
+        if _, interesting := tablesOfInterest[table]; interesting {
+            var columns []string
+            for column := range tableSchema.Columns {
+                columns = append(columns, column)
+            }
+            monitorRequests[table] = libovsdb.MonitorRequest{
+                Columns: columns,
+                Select: libovsdb.MonitorSelect{
```
Comment (vishpat) [inline api/Port.go]: Not sure if we need to monitor all of these. Modify must be the only one we need to monitor

File: api/Port.go [hunk 7]
```diff
@@ -151,3 +194,168 @@ func (vrsConnection *VRSConnection) UpdatePortMetadata(name string, metadata map
...
+                empty := libovsdb.Row{}
+                clientEvent := &ClientEvent{}
+                if !reflect.DeepEqual(row.New, empty) {
+                    clientEvent = ovsdbc.CreateObject(table, row.New, Add)
+                } else {
+                    clientEvent = ovsdbc.CreateObject(table, row.Old, Delete)
+                }
+                if clientEvent != nil {
+                    log.Debugf("Writing th ovsdb entry into clientChannel for consumption", *clientEvent)
+                    clientChan <- *clientEvent
```
Comment (vishpat) [inline api/Port.go]: Why are we doing this ?

Comment (vishpat) [reply api/Port.go]: If the ip information is available at this point it makes sense. And if the ip is available, we can just exist the go routine after this.

File: api/Port.go [hunk 8]
```diff
@@ -151,3 +194,168 @@ func (vrsConnection *VRSConnection) UpdatePortMetadata(name string, metadata map
...
+                            log.Debugf("Writing th ovsdb entry into clientChannel for consumption")
+                            clientChan <- *clientEvent
+                        } else {
+                            log.Infof("ClientEvent object is empty for OVSDB Nuage_Port_Table update")
+                        }
+                    }
+                }
+
+
+            case <-ovsdbc.StopChannel:
```
Comment (vishpat) [inline api/Port.go]: Remove

File: api/NuageAPI_test.go [hunk 9]
```diff
@@ -305,25 +310,25 @@ func TestVMCreateDelete(t *testing.T) {
 	t.Logf("VM %s got resolved with an IP address %s on VRS", vmInfo["name"], portIPOnVRS)
 
 	// Comparing port's IP address on VRS and VSD
-	if portIPOnVSD != portIPOnVRS {
-		t.Fatal("Port IPs on VSD and VRS do not match.")
-	} else {
-		t.Logf("Port IPs on VSD and VRS match.")
-	}
+	//if portIPOnVSD != portIPOnVRS {
```
Comment (vishpat) [inline api/NuageAPI_test.go]: Remove

File: api/NuageAPI_test.go [hunk 10]
```diff
@@ -305,25 +310,25 @@ func TestVMCreateDelete(t *testing.T) {
...
 
 	// Performing cleanup of port/entity on VRS
 	err = cleanup(vrsConnection, vmInfo)
 	if err != nil {
 		t.Fatal("Unable to delete port from OVSDB table")
 	}
 
-	t.Logf("Waiting for 300 seconds before verifying port gets removed from VRS")
-	time.Sleep(time.Duration(300) * time.Second)
+	//t.Logf("Waiting for 300 seconds before verifying port gets removed from VRS")
```
Comment (vishpat) [inline api/NuageAPI_test.go]: remove

File: api/NuageAPI_test.go [hunk 11]
```diff
@@ -278,16 +279,20 @@ func TestVMCreateDelete(t *testing.T) {
...
 	}
 
-	t.Logf("Waiting for 60 seconds before verifying port got resolved with an IP address in VRS")
-	time.Sleep(time.Duration(60) * time.Second)
+	//t.Logf("Waiting for 60 seconds before verifying port got resolved with an IP address in VRS")
+	//time.Sleep(time.Duration(60) * time.Second)
+
+	// Registering for OVSDB updates instead of random sleep
+	ovsdbUpdateChan := client.GetNuagePortTableUpdate()
+    	fmt.Println(<-ovsdbUpdateChan)
```
Comment (vishpat) [inline api/NuageAPI_test.go]: Can you grab the ip here and use it for verification below

File: api/Port.go [hunk 12]
```diff
@@ -5,8 +5,52 @@ import (
...
+
+// ClientEvent defines object type returned
+// from OVSDB monitoring
+type ClientEvent struct {
+	EventData interface{}
+}
+
+// OvsdbClient defines OVSDB client to register
+// and monitor OVSDB table updates
+type OvsdbClient struct {
```
Comment (vishpat) [inline api/Port.go]: Get rid of this

File: api/Port.go [hunk 13]
```diff
@@ -5,8 +5,52 @@ import (
 	"github.com/nuagenetworks/libvrsdk/api/port"
 	"github.com/nuagenetworks/libvrsdk/ovsdb"
 	"github.com/socketplane/libovsdb"
+	"reflect"
+	"time"
 )
 
+// Empty structure.
+type Empty struct{}
```
Comment (vishpat) [inline api/Port.go]: lowercase

File: api/Port.go [hunk 14]
```diff
@@ -5,8 +5,52 @@ import (
...
+
+// OvsdbEventType defines event type sent to OVSDB
+type OvsdbEventType string
```
Comment (vishpat) [inline api/Port.go]: lowercase

File: api/Port.go [hunk 15]
```diff
@@ -5,8 +5,52 @@ import (
...
+
+// Const definitions for OVSDB event types
+const (
```
Comment (vishpat) [inline api/Port.go:15]: lowercase

File: api/Port.go [hunk 16]
```diff
@@ -5,8 +5,52 @@ import (
...
+	Add    OvsdbEventType = "ADD"
+	Update OvsdbEventType = "UPDATE"
+)
+
+// ClientEvent defines object type returned
+// from OVSDB monitoring
+type ClientEvent struct {
```
Comment (vishpat) [inline api/Port.go]: lowercase

File: api/Port.go [hunk 17]
```diff
@@ -5,8 +5,52 @@ import (
...
+// for container port resolved in OVSDB
+type PortIPv4Info struct {
+	ipaddr  string
+	gateway string
+	mask    string
+}
+
+// OvsdbEvent defines structure for
+// OVSDB event
+type OvsdbEvent struct {
```
Comment (vishpat) [inline api/Port.go]: lowercase

File: api/Port.go [hunk 18]
```diff
@@ -151,3 +195,149 @@ func (vrsConnection *VRSConnection) UpdatePortMetadata(name string, metadata map
...
+
+// Update will provide updates on OVSDB table updates
+func (ovsdbc OvsdbClient) Update(context interface{}, tableUpdates libovsdb.TableUpdates) {
+	ovsdbc.mUpdateChan <- &tableUpdates
+}
+
+// GetNuagePortTableUpdate will register with OVSDB
+// for Nuage Port table updates and return as soon as
+// port table entry gets populated
+func (ovsdbc *OvsdbClient) GetNuagePortTableUpdate() <-chan ClientEvent {
```
Comment (vishpat) [inline api/Port.go]: Please define this method on VRSConnection struct

File: api/Port.go [hunk 19]
```diff
@@ -151,3 +195,149 @@ func (vrsConnection *VRSConnection) UpdatePortMetadata(name string, metadata map
...
+	var err error
+	var conn *libovsdb.OvsdbClient
+	ovsdbc.mUpdateChan = make(chan *libovsdb.TableUpdates)
+	clientChan := make(chan ClientEvent, 1)
+
+        for {
```
Comment (vishpat) [inline api/Port.go]: Please remove

File: api/Port.go [hunk 20]
```diff
@@ -151,3 +195,149 @@ func (vrsConnection *VRSConnection) UpdatePortMetadata(name string, metadata map
...
+                if err != nil {
+                        //log.Infof("Couldn't connect to ovsdb server while trying to register for Nuage Port table updates")
+                } else {
+                        break
+                }
+                time.Sleep(time.Second * time.Duration(5))
+        }
+
+        ovsdbc.mConn = conn
+        ovsdbc.mConn.Register(ovsdbc)
```
Comment (vishpat) [inline api/Port.go]: I had mentioned before, conn.Register(ovsdbc)

File: api/Port.go [hunk 21]
```diff
@@ -151,3 +195,149 @@ func (vrsConnection *VRSConnection) UpdatePortMetadata(name string, metadata map
...
+		if _, interesting := tablesOfInterest[table]; interesting {
+			var columns []string
+			for column := range tableSchema.Columns {
+				columns = append(columns, column)
+			}
+			monitorRequests[table] = libovsdb.MonitorRequest{
+				Columns: columns,
+				Select: libovsdb.MonitorSelect{
+					Initial: true,
+					Insert:  true,
```
Comment (vishpat) [inline api/Port.go]: Why are monitoring inserts and deletes ?

File: api/Port.go [hunk 22]
```diff
@@ -151,3 +195,149 @@ func (vrsConnection *VRSConnection) UpdatePortMetadata(name string, metadata map
...
+					Delete:  true,
+					Modify:  true}}
+		}
+	}
+	initialData, err := ovsdbc.mConn.Monitor("Open_vSwitch", nil, monitorRequests)
+	if err != nil {
+		//log.Errorf("Unable to register for OVSDB notifications %v", err)
```
Comment (vishpat) [inline api/Port.go]: We cannot ignore errors. This needs to be handled, push empty data on the channel.

File: api/Port.go [hunk 23]
```diff
@@ -151,3 +195,149 @@ func (vrsConnection *VRSConnection) UpdatePortMetadata(name string, metadata map
...
+	go func() {
+		for table, tableUpdate := range initialData.Updates {
+			for _, row := range tableUpdate.Rows {
+				empty := libovsdb.Row{}
+				clientEvent := &ClientEvent{}
+				if !reflect.DeepEqual(row.New, empty) {
+					clientEvent = ovsdbc.CreateObject(table, row.New, Add)
+				}
+				if clientEvent != nil {
+					//log.Debugf("Writing th ovsdb entry into clientChannel for consumption", *clientEvent)
```
Comment (vishpat) [inline api/Port.go]: What if clientEvent is valid ? i.e. the ip got resolved by the time this API was called. Why is this being ignored ?

File: api/Port.go [hunk 24]
```diff
@@ -151,3 +195,149 @@ func (vrsConnection *VRSConnection) UpdatePortMetadata(name string, metadata map
...
+					//clientChan <- *clientEvent
+				} else {
+					//log.Debugf("ClientEvent object is empty for OVSDB Nuage_Port_Table update")
+				}
+			}
+		}
+		for {
+			currentUpdate := <-ovsdbc.mUpdateChan
+			for table, tableUpdate := range currentUpdate.Updates {
```
Comment (vishpat) [inline api/Port.go]: This for loop looks quite similar to the above one. Create a common function to encapsulate the functionality of the for loop and use it at both the places.

File: api/Port.go [hunk 25]
```diff
@@ -151,3 +195,149 @@ func (vrsConnection *VRSConnection) UpdatePortMetadata(name string, metadata map
...
+	}()
+	return clientChan
+}
+
+// CreateObject will create an object filled with
+// required fields from OVSDB tables
+func (ovsdbc *OvsdbClient) CreateObject(table string, row libovsdb.Row, ovsdbEventType OvsdbEventType) *ClientEvent {
+	switch table {
+	case "Nuage_Port_Table":
+		ovsdbObject := PortIPv4Info{}
```
Comment (vishpat) [inline api/Port.go]: Rename ovsdbObject to portIPv4Info

File: api/Port.go (hunk 18 above)
Comment (rparulek) [reply api/Port.go]: Rename to GetPortIPv4Info

File: api/NuageAPI_test.go [hunk 26]
```diff
@@ -278,8 +278,9 @@ func TestVMCreateDelete(t *testing.T) {
 		t.Fatal("Unable to create a test VM")
 	}
 
-	t.Logf("Waiting for 60 seconds before verifying port got resolved with an IP address in VRS")
-	time.Sleep(time.Duration(60) * time.Second)
+	// Registering for OVSDB updates instead of random sleep
+	ovsdbUpdateChan := vrsConnection.GetPortIPv4Info()
```
Comment (vishpat) [inline api/NuageAPI_test.go]: rename this to portInfoUpdateChan

File: api/NuageAPI_test.go [hunk 27]
```diff
@@ -595,8 +585,9 @@ func TestVMHotNICAdd(t *testing.T) {
 		t.Fatal("Unable to add Port to entity")
 	}
 
-	t.Logf("Waiting for 60 seconds before verifying new port added to an existing entity gets resolved")
-	time.Sleep(time.Duration(60) * time.Second)
+	// Registering for OVSDB updates instead of random sleep
+        ovsdbUpdateChan = vrsConnection.GetPortIPv4Info()
```
Comment (vishpat) [inline api/NuageAPI_test.go]: Same as above

File: api/Port.go [hunk 28]
```diff
@@ -151,3 +175,104 @@ func (vrsConnection *VRSConnection) UpdatePortMetadata(name string, metadata map
...
+					Modify:  true}}
+		}
+	}
+	initialData, err := vrsConnection.ovsdbClient.Monitor("Open_vSwitch", nil, monitorRequests)
+	if err != nil {
+		return clientChan
+	}
+
+        portInfo := PortIPv4Info{}
+        addedTable, addedRow := getUpdatedPortTableRow(initialData, add)
```
Comment (vishpat) [inline api/Port.go]: Rename addedTable to added

File: api/Port.go [hunk 29]
```diff
@@ -151,3 +175,104 @@ func (vrsConnection *VRSConnection) UpdatePortMetadata(name string, metadata map
...
+        if addedTable == true { 
+                portInfo = getPortInfo(porttable, addedRow, add)
+        }
+        if portInfo.ipaddr != "" && portInfo.gateway != "" && portInfo.mask != "" {
+        	clientChan <- portInfo
+		return clientChan
```
Comment (vishpat) [inline api/Port.go:194]: Don't return channel here, go routines don't return anything

Comment (vishpat) [reply api/Port.go:194]: Add simple return

File: api/Port.go [hunk 30]
```diff
@@ -151,3 +175,104 @@ func (vrsConnection *VRSConnection) UpdatePortMetadata(name string, metadata map
...
+        }
+
+	go func() {
+                for {
+                        currentUpdate := <-vrsConnection.updatesChan
+                        updatedTable, updatedRow := getUpdatedPortTableRow(currentUpdate, update)
```
Comment (vishpat) [inline api/Port.go]: rename updatedTable to updated

File: api/Port.go [hunk 31]
```diff
@@ -151,3 +175,104 @@ func (vrsConnection *VRSConnection) UpdatePortMetadata(name string, metadata map
...
+                        if updatedTable == true {
```
Comment (vishpat) [inline api/Port.go]: What if updatedTable is false ? Does that imply that the update is for a port in which we are not interested ?

File: api/Port.go [hunk 32]
```diff
@@ -151,3 +175,104 @@ func (vrsConnection *VRSConnection) UpdatePortMetadata(name string, metadata map
...
+		}
+	}
+	initialData, err := vrsConnection.ovsdbClient.Monitor("Open_vSwitch", nil, monitorRequests)
+	if err != nil {
+		return clientChan
+	}
+
+        portInfo := PortIPv4Info{}
+        addedTable, addedRow := getUpdatedPortTableRow(initialData, add)
+        if addedTable == true {
```
Comment (vishpat) [inline api/Port.go]: What if addedTable is false ?  Does that imply that the ovsdb row is for a port in which we are not interested ?

File: api/Port.go [hunk 33]
```diff
@@ -151,3 +175,104 @@ func (vrsConnection *VRSConnection) UpdatePortMetadata(name string, metadata map
...
+                        }
+                	if portInfo.ipaddr != "" && portInfo.gateway != "" && portInfo.mask != "" {
+                                clientChan <- portInfo
+                        }
+                }
+        }()
+        return clientChan
+}
+
+func getUpdatedPortTableRow(data *libovsdb.TableUpdates, ovsdbEventType ovsdbEventType) (bool,libovsdb.Row) {
```
Comment (vishpat) [inline api/Port.go]: Rename to getAddedOrUpdatedPortTableRow

File: api/Port.go [hunk 34]
```diff
@@ -151,3 +175,104 @@ func (vrsConnection *VRSConnection) UpdatePortMetadata(name string, metadata map
...
+
+        for _, tableUpdate := range data.Updates {
+                for _, row := range tableUpdate.Rows {
+                        empty := libovsdb.Row{}
+                        if !reflect.DeepEqual(row.New, empty) && ovsdbEventType == add {
+                                return true, row.New
+                        }
+
+                        if !reflect.DeepEqual(row.New, empty) && !reflect.DeepEqual(row.Old, empty) && ovsdbEventType == update {
```
Comment (vishpat) [inline api/Port.go]: Should this be row.Updated ?

File: api/Port.go (hunk 29 above)
Comment (rparulek) [reply api/Port.go:194]: But isn't this outside the goroutine?

File: api/Port.go (hunk 31 above)
Comment (rparulek) [reply api/Port.go]: updatedTable is false implies that the table did not get updated

File: api/Port.go [hunk 35]
```diff
@@ -151,3 +175,104 @@ func (vrsConnection *VRSConnection) UpdatePortMetadata(name string, metadata map
...
+                                return true, row.New
+                        }
+                }
+        }
+
+        return false, libovsdb.Row{}
+}
+
+func getPortInfo(table string, row libovsdb.Row, ovsdbEventType ovsdbEventType) PortIPv4Info {
```
Comment (vishpat) [inline api/Port.go:232]: Rename to extractPortInfoFromRow

**Pull request closed without merging**